dbms/src/DataStreams/ParallelInputsProcessor.h

```diff
@@ -237,127 +283,81 @@ class ParallelInputsProcessor
         {
             handler.onException(exception, thread_num);
         }
+    }
 
-        handler.onFinishThread(thread_num);
+    /// This function may be called in different threads.
+    /// If finish is not true and no exception occurs, we can ensure that the work is
+    /// all done when the function returns in any thread.
+    void loop(size_t thread_num, WorkingInputs & work)
+    {
+        if (work.active_inputs == 0)
+        {
+            return;
+        }
 
-        /// The last thread on the output indicates that there is no more data.
-        if (0 == --active_threads)
+        while (!finish)
```

Comment:
After using `MPMCQueue` you needn't depend on `finish`, the queue itself contains enough information.